File: Gen_ADC_Data/src/SAR_ADC.cpp

#include <iostream>
#include <vector>
#include <cmath>
#include <random>
#include "SAR_ADC.h"

using namespace std;

SAR_ADC::SAR_ADC(int nbit){
	Nbit = nbit;
	codes = (int)pow(2.0,Nbit);	

	LSB = 0.0;
	Vsig = 0.0;
	Voff = 0.0;
	Cpar = 0.0;
	
	for(int i=0;i<Nbit-1;i++){
		VDAC.push_back(0.0);
		Cmis.push_back(0.0);
		double cap = pow(2.0,i);
		Cap.push_back(cap);
	}
	for(int i=0;i<codes;i++){
		INL.push_back(0.0);
		DNL.push_back(0.0);
		Up_Limit.push_back(0.0);
	}
}


int  SAR_ADC::sort_Code(int icode){
	int num = icode;
	int bit0 = num%2;
	for(int i=0;i<Nbit;i++){
		if(num%2 != bit0) return i-1;
		num = (int)(num/2);
	}
}

void SAR_ADC::set_Cap_Mismatch(vector<double>& cmis){
	if(cmis.size()<Nbit-1){
		std::cout<<"Error: The input Mismatch array is too short";
		return;
	}
	for(int i=0;i<Nbit-1;i++){
		Cmis[i] = cmis[i];
	}
}

void SAR_ADC::set_Offset(double voff){
	Voff = voff;
}

void SAR_ADC::set_Parasitic(double cpar){
	Cpar = cpar;
}

void SAR_ADC::set_Noise_Sigma(double vsig){
	Vsig = vsig;
}

void SAR_ADC::gen_VDAC_and_LSB(){
	double Csum = Cpar + 1.0;
	for(int i=0;i<Nbit-1;i++) Csum += Cap[i] + Cmis[i];
	double VDAC_sum = 0.0;
	for(int i=0;i<Nbit-1;i++){
		VDAC[i] = 1.0*(Cap[i]+Cmis[i])/Csum;
		VDAC_sum += VDAC[i];
	}
	LSB = 2.0*VDAC_sum/(codes-2);
}

void SAR_ADC::gen_DNL_and_INL(){
	vector<double> delta_VDAC;
	for(int i=0;i<Nbit-1;i++) delta_VDAC.push_back(0.0);
	delta_VDAC[0] = VDAC[0];
	double delta_sum = 0.0;
	for(int i=1;i<Nbit-1;i++){
		delta_VDAC[i] = VDAC[i] - delta_sum;
		delta_sum += VDAC[i];
		// for Debug
		std::cout<<"delta VDAC["<<i<<"]="<<delta_VDAC[i]<<endl;
	}
	DNL[0] = 0.0; DNL[codes-1] = 0.0;
	for(int i=1;i<codes-1;i++){
		double sort_code = sort_Code(i);
		DNL[i] = 1.0*delta_VDAC[sort_code]/LSB - 1.0;
	}
	INL[0] = 0.0;
	for(int i=1;i<codes-1;i++){
		INL[i] = INL[i-1] + DNL[i-1];
	}
}

int  SAR_ADC::gen_Code_and_Vres_without_noise(double vi, double& vres){
	vector<double> bit;
	for(int i=0;i<Nbit;i++) bit.push_back(0);
	vres = vi;
	for(int i=0;i<Nbit-1;i++){
		if(vres>Voff) bit[i] = 1;
		vres -= (2*bit[i]-1)*VDAC[Nbit-2-i];
	}
	if(vres>Voff) bit[Nbit-1] = 1;
	int value = 0;
	for(int i=0;i<Nbit;i++) value += pow(2.0,Nbit-1-i)*bit[i];
	return (int)value;
	
}

int  SAR_ADC::gen_Code_and_Vres_with_noise(double vi, double& vres){
	std::random_device rdev{};
	std::default_random_engine generator{rdev()};
	std::normal_distribution<double> distribution(0.0,Vsig);

	vector<double> bit;
	for(int i=0;i<Nbit;i++) bit.push_back(0);
	vres = vi;
	double noise; 
	for(int i=0;i<Nbit-1;i++){
		noise = distribution(generator);
		vres = vres + noise;
		if(vres > Voff) bit[i] = 1;
		vres -= (2*bit[i]-1)*VDAC[Nbit-2-i];
	}
	noise = distribution(generator);
	vres = vres + noise;
	if(vres>Voff) bit[Nbit-1] = 1;
	int value = 0;
	for(int i=0;i<Nbit;i++) value += pow(2.0,Nbit-1-i)*bit[i];
	return (int)value;
	
}

double SAR_ADC::get_Up_Limit_for_Code(int icode){
	vector<int> bits;
	int num = icode;
	int bit;
	for(int i=0;i<Nbit;i++){
		bit = num%2;
		bits.push_back(bit);
		num = (int)(num/2);
	}
	double res = 0.0, up = 2.0, dn = -2.0;
	for(int i=Nbit-1;i>=1;i--){
		if(bits[i]>0){
			dn = (dn>res)?dn:res;
			res += VDAC[i-1];
		}
		else{
			up = (up<res)?up:res;
			res -= VDAC[i-1];
		}
	}
	if(bits[0]<0) up = (up<res)?up:res;
	if(bits[0]>0) dn = (dn>res)?dn:res;

	if(up<dn) return -2;	// In this case, this code is missing!
	else return up;
}

void SAR_ADC::get_Up_Limit(){
	for(int i=0;i<codes-1;i++){
		double up = get_Up_Limit_for_Code(i);
		Up_Limit[i] = up;
	}
	Up_Limit[codes-1] = 1.0;
}

File: python/SAR_ADC.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
This is the definition of binary MCS SAR ADC.

Please Keep In Mind: [0] represents the LSB!

Author: Zhen-Xiong Yuan
"""

import math
import random

class SAR_ADC(object):
    def __init__(self, Nbit):
        self.Nbit = Nbit
        self.codes = int(math.pow(2,Nbit))
        self.Cap = [math.pow(2,i) for i in range(Nbit-1)]
        self.Cmis = [0.0*i for i in range(Nbit-1)]
        self.Cpar = 0.0
        self.Voff = 0.0
        self.Vsig = 0.0

        self.VDAC = [0.0*i for i in range(Nbit-1)]
        self.DNL = [0.0*i for i in range(self.codes)]
        self.INL = [0.0*i for i in range(self.codes)]
        self.LSB = 0.0

    def set_Cap_Mismatch(self, Cmis):
        self.Cmis = Cmis

    def set_Offset(self, Voff):
        self.Voff = Voff

    def set_Parasitic(self, Cpar):
        self.Cpar = Cpar

    def set_Noise_Sigma(self, Vsig):
        self.Vsig = Vsig

    def gen_VDAC_and_LSB(self):
        Csum = self.Cpar + 1.0
        for i in range(self.Nbit-1): Csum += self.Cap[i] + self.Cmis[i]
        VDAC_sum = 0.0
        for i in range(self.Nbit-1):
            self.VDAC[i] = 1.0*(self.Cap[i] + self.Cmis[i])/Csum
            VDAC_sum += self.VDAC[i]
    
        self.LSB = 2.0*VDAC_sum/(self.codes - 2)

    def sort_Code(self,code):
        num = int(code)
        bit0 = num%2
        for i in range(self.Nbit):
            if num%2 != bit0:
                return i-1
            num = int(num/2)
        
    def gen_DNL(self):
        delta_VDAC = [0.0*i for i in range(self.Nbit-1)]
        delta_VDAC[0] = self.VDAC[0]
        delta_sum = delta_VDAC[0]
        for i in range(1,self.Nbit-1):
            delta_VDAC[i] = self.VDAC[i] - delta_sum
            delta_sum += self.VDAC[i]
            print("delta_VDAC[",i,"]=",delta_VDAC[i])
        self.DNL[0] = 0.0
        self.DNL[self.codes-1] = 0.0
        for i in range(1,self.codes-1):
            sort_code = self.sort_Code(i)
            self.DNL[i] = 1.0*delta_VDAC[sort_code]/self.LSB - 1.0

    def gen_INL(self):
        self.INL[0] = 0.0
        for i in range(1,self.codes):
            self.INL[i] = self.INL[i-1] + self.DNL[i-1]

    def gen_Code_and_Vres_Without_Noise(self, vi):
        code = [0*i for i in range(self.Nbit)] # keep in mind MSB is [0]
        vres = vi
        for i in range(self.Nbit-1):
            if (vres>self.Voff): code[i]=1
            vres -= (2*code[i]-1)*self.VDAC[self.Nbit-2-i]
        if (vres>self.Voff): code[self.Nbit-1] = 1
        value = 0
        for i in range(self.Nbit): 
            value += math.pow(2,self.Nbit-1-i)*code[i]
        return int(value), vres 

    def gen_Code_and_Vres_With_Noise(self, vi):
        code = [0*i for i in range(self.Nbit)] # keep in mind MSB is [0]
        vres = vi
        for i in range(self.Nbit-1):
            if (vres>self.Voff+random.gauss(0.0,self.Vsig)): code[i]=1
            vres -= (2*code[i]-1)*self.VDAC[self.Nbit-2-i]
        if (vres>self.Voff+random.gauss(0.0,self.Vsig)): 
            code[self.Nbit-1] = 1
        value = 0
        for i in range(self.Nbit): 
            value += math.pow(2,self.Nbit-1-i)*code[i]
        return int(value), vres 





















